### PR TITLE
Reject failed inits

### PR DESCRIPTION
**Pull request by ankitson**
This is a fix for #4416 . Its still WIP, some cleanup & style fixes needed I think, but the basic idea is here so I thought I would open it up to suggestions.

@allanrenucci It mostly does what you suggested, except we also have to patch `compile` to account for failures in `displayDefinitions`. 

It doesn't seem very clean, but it is explicit about the dependency between `display` and the REPL state. I think it will remain a little messy until we decouple eval and rendering.

### CODE REVIEW

**Comment by ankitson** [comment]
Also found another weird behavior:

```scala> lazy val x = 1/0
val x: Int = <lazy>
scala> val y = x
java.lang.ArithmeticException: / by zero
	...
scala> y
1 |y
  |^
  |not found: y
scala> x
<rendering error>
```

And without this patch the REPL reports `x=0`.

```scala> lazy val x = 1/0
val x: Int = <lazy>
scala> val y = x
java.lang.ArithmeticException: / by zero
	...
scala> y
1 |y
  |^
  |not found: y
scala> x
val res1: Int = 0
```

File: compiler/src/dotty/tools/repl/Rendering.scala [hunk 1]
```diff
@@ -70,21 +68,20 @@ private[repl] class Rendering(compiler: ReplCompiler,
...
 
-    try {
+    Try {
       val resultValue =
         if (d.symbol.is(Flags.Lazy)) Some("<lazy>")
         else valueOf(d.symbol)
 
       resultValue.map(value => s"$dcl = $value")
     }
-    catch { case ex: InvocationTargetException => Some(renderError(ex)) }
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/repl/Rendering.scala]: We should propagate any exception that is not an `InvocationTargetException`

File: compiler/src/dotty/tools/repl/ReplDriver.scala [hunk 2]
```diff
@@ -279,15 +280,20 @@ class ReplDriver(settings: Array[String],
...
         info.bounds.hi.typeMembers.filter(_.symbol.info.isInstanceOf[TypeAlias])
 
-      (
-        typeAliases.map("// defined alias " + _.symbol.showUser) ++
-        defs.map(rendering.renderMethod) ++
-        vals.map(rendering.renderVal).flatten
-      ).foreach(str => out.println(SyntaxHighlighting(str)))
-
-      state.copy(valIndex = state.valIndex - vals.filter(resAndUnit).length)
+      val (successes, failures) = vals.map(rendering.renderVal).partition(_.isSuccess)
```
[the remark this answers is not in the document]
Comment (allanrenucci) [reply compiler/src/dotty/tools/repl/ReplDriver.scala]: And not render anything (e.g. aliases, methods) in case of failure

**Comment by ankitson** [comment]
@allanrenucci I managed to get this working. For some reason I don't fully understand, the `Rendering` class gets into a bad state if we throw an exception while evaluating, this happens:

```
scala> val x = 1/0
java.lang.ArithmeticException: / by zero
...
scala> val y = 1/0

scala> def f = y
def f: Int
```

My PR now creates a new `Rendering` upon failure.

**Comment by OlivierBlanvillain** [comment]
@ankitson Sorry for the long silence. Do still plan to work on this PR? The latest commits seems like a hack rather than a clean solution; a `var` with a `scala.util.control.Breaks.{break, breakable}`, we are not going to merge a diff like that...

**Comment by OlivierBlanvillain** [comment]
Closing for now, feel free to reopen if you plan to further work on it!

**Pull request closed without merging**